### ISSUE TITLE
Consider using WEAT names as a test case for countering bias

### ISSUE THREAD

**Issue by rspeer** [conversationai/perspectiveapi#22]
I've been one of many to criticize the deployed PerspectiveAPI system, and I got a response from @iislucas on Twitter pointing to research slides about de-biasing PerspectiveAPI. I'm glad to see it's something you're working on. And yes, it's a positive thing about PerspectiveAPI that its results (though not its code and data) are open for the world to test -- it's certainly a better situation than having biased algorithms that operate in secret.

I notice in the slides that you discuss various approaches to de-biasing NLP, but end up kind of dismissing many of them in favor of Google's own solution. I think that spending a bit more time with other de-biasing research would help you discover the problems with PerspectiveAPI and why everyone is so angry about it, and could even provide test cases that let you get ahead of the issue.

One example that I think should become part of your test suite is Caliskan et al.'s Word Embedding Association Test (WEAT). Although iislucas tells me that, contrary to my intuition, PerspectiveAPI's biases don't come from word embeddings, the test data comparing predominantly-white to predominantly-black names can still be used on the output of PerspectiveAPI.

I've done this using the demo at perspectiveapi.com. I entered the sentence "Hello, my name is ____" while filling each name from the WEAT 4 lists into the blank. The results show a highly statistically significant bias against African-Americans:

| | African-American names | European-American names |
| --- | --- | --- |
| N | 17 | 18 |
| mean | .1118 | .0278 |
| stdev | .0689 | .0163 |
| SEM | .0167 | .0038 |

(Incidentally, N=17 for the African-American names because the API refuses to categorize the sentence "Hello, my name is Latonya", judging it to be non-English.)

To be clear: I think the wrong response would be to put sentences like this into the training data like in #19. This test is _evidence_ that the algorithm is biased, and training on it would remove the evidence but not the bias.

I also remain puzzled by the assertion that this isn't word-embedding bias. The pattern of bias against certain names strongly resembles the pattern in uncorrected word2vec, and if it's not that, why would the model consider it so strongly distinguishing to be named Jermaine versus being named Greg?

So my suggestion here is that you test PerspectiveAPI on the WEAT lists to help identify what's not working about de-biasing. I'd encourage adding the other comparisons, such as male vs. female and young vs. old. This will help to deal with this issue in development instead of the Internet at large discovering it in production.

**Comment by cjadams** [comment]
Thank you for the suggestion. Will investigate if this method could be added to model bias evaluation process.

**Comment by iislucas** [comment]
Thanks for the suggestions @rspeer ! 
 
The names discrepancy in scores has a rather boring reason: most unusual names are unknown tokens, and unknown tokens along naturally have a slightly higher score. But this does of course lead to various interesting approaches to both measuring bias, as you suggest, and mitigating it, that we've been exploring. For example, ablation using name lists to make all names unknown tokens. These approaches of course also have their problems; for example  ablation means you loose information about who's talking about who (and yes, you can get more complex here too, like using debruijn index as names, etc). Other approaches include adversarial training, tinkering with loss functions, fuzzing original datasets etc. Lots of interesting areas to explore. If you want to try some yourself, you can find public code for models here: https://github.com/conversationai/unintended-ml-bias-analysis/blob/master/unintended_ml_bias/model_tool.py . And we created a recent Kaggle competition which provides a new public dataset also here: kaggle.com/c/jigsaw-toxic-comment-classification-challenge We also made a public dataset with demographic information of raters about a year ago here: https://figshare.com/articles/Wikipedia_Talk_Labels_Toxicity/4563973 
The challenge really is not what can be tried, but which of the many directions to go in is most likely to work; like many ares of science, each potential answer opens up a huge array of new questions!

W.r.t. Your proposal not to mix training/test data is of course right and we agree. And just in case it's not clear, that is *not* what was done in #19 What we did was create an extended test set and training set. You can read more about what we did, and how it's related to other approaches in our recent paper: http://www.aies-conference.com/wp-content/papers/main/AIES_2018_paper_9.pdf (Published at the AIES'2017 conference: http://www.aies-conference.com/)

W.r.t. word embedding bias; because word-embeddings are only used as an initialization word vector, and then back-prop happens through them, the starting word-embeddings on deep models actually doesn't make a lot of difference once you have enough data. There's a recent paper that talks a bit about this here: https://arxiv.org/pdf/1711.09050.pdf

There is a lot of other great work on the challenges of unintended bias, fairness, and how to tackle it in the FATML/FAT* community (https://www.fatml.org/) as well as many other places, you can find more of our own analysis and links to related work here: https://conversationai.github.io/bias

**Comment by rspeer** [comment]
I'm fairly up-to-date on the research and I followed the FAT* conference as much as I could over the livestream. And that's an informative reference showing that reducing word-embedding bias is not sufficient. (It may still be _necessary_.)

The research shows your good intentions, but I feel that a lot of this is distracting from the real fact that this unfortunately biased system has been in production for over a year, and people are unquestioningly using its results. This is quite a productive research topic, but it's not _just_ a research topic.

The AIES paper is interesting but the results seem to show the limitations of what can be done with synthetic changes to the input. Interventions in the model itself may be necessary.

I don't see how this Kaggle competition relates to de-biasing. It appears to not measure bias, only accuracy. I suppose the test set could have identity-related words shuffled around compared to the training set, but I doubt you're doing that unannounced.

**Comment by iislucas** [comment]
You can start from randomized word embeddings and find yourself still generating biased results - so yes, it's clearly not sufficient; it may be that much further down that road, the effect of word-embedding bias can generate a measurable improvement w.r.t. unintended biases, but I'm not confident its a very likely to be helpful. I think other approaches like the ones I mentioned are much likelier to lead to more significant improvements. But it's a big open question what methods help how much. There is no silver bullet here.

W.r.t. the question of "should one use a model that has unintended biases", I think if your answer to that is "no", then you basically stop yourself from experimenting and trying to make the world better with ML. I don't know of an ML system trained on real world data *without* unintended biases (every sentiment model I know of has them, as do image classifiers, translation systems, etc.).  So I think the more important question is not about ML's unintended biases, but about the significance of the unintended biases in real use cases. For that you have to look at how the system is actually used. For example, what is the human process that interacts with the model, and what thresholds are used, and what happens to the resulting artifacts, and how often does the bias result in the model making an incorrect judgment, and how incorrect is it, and what affect does that then have? These, I think, are the really important questions to look at for applied systems. In general, the fairness-impact of a model is actually orthogonal to the bias the model has, and even fictional fair models can be used in unfair ways. Research into making models fairer is super important, but its certainly also not enough.

We did some experiments using imperfect models on real data and wrote a short blog post about one example here: https://medium.com/the-false-positive/better-discussions-with-imperfect-models-91558235d442 
In the use cases that I've looked at, it looks to me like even the groups affected by the ML bias of these text classifiers benefit from the models impact on conversation (perhaps even disproportionately positively). The way we tackle understanding the challenge is to work with impact groups (like [GLAAD](https://www.glaad.org/blog/glaad-launches-digital-task-force-advocacy-groups-tech-companies-and-community-leaders-address)), but even then it's hard to get more than anecdotal data, at the moment this seems about as good a way to understand real impact as I can see. Of course, if we see other evidence, we'll change direction and recommendations accordingly. But right now it seems to me these tools, even in their early stages, are making conversations more inclusive, even to the groups who's names the models tends to give higher scores to, and to even in the other cases where I consider the scores to be frequently problematic. One positive observation here is that every use case I know about has followed our proposal of keeping a human in the loop, and using models to help people facilitate online conversations rather than try to automate what is shown (you might notice that our public demos were focused on authorship and viewership experiences, this is not accidental). This suggests that people using APIs like Perspective are doing a fairly reasonable job of applying it. We spend a fair bit of time trying to understand the actual use-cases, and check this is the case, but if you see a use case where the overall impact is negative to our goals of increasing participation, quality and empathy in discussion, do file a issue to we can explore it. 

The Kaggle competition, and associated Figshare data on demographics that I linked to, provides data, code, and models for you (and anyone else interested) to experiment and explore the nature of unintended bias with a permissive license to do so. Our [the github project with our code and paper](https://github.com/conversationai/unintended-ml-bias-analysis/blob/master/unintended_ml_bias) in particular provides tools and new approaches to measure impact of approaches to tackle unintended bias (which I think is important, as when we started this work, the state of metrics was frustratingly unfitting for text classifiers). The approach we took is actually quite close to your suggestion for names, but based on AUC scores and multiple subsets. The wider goal is to help make online conversations better, and do to so according to [our values](https://conversationai.github.io/#values-guiding-the-research).

**Comment by rspeer** [comment]
First of all, I'm glad that you're willing to discuss this. I certainly understand more about how the model got this way than I did in my original post.

I don't mean to say that you shouldn't experiment, but I am asking you to fix one of the bad effects of the experiment, which is really what this issue report is about.

I am unconvinced by the blog post about "Better discussions with imperfect models". Warning users that the model can make incorrect classifications just sounds like a generic, obvious warning about ML. It sounds like something your legal department made you put there, and it doesn't come with any context about what those incorrect classifications are.

I agree with the blog post's point that it's useful to be able to find the toxic comments in a large list more easily, and that even a flawed classifier can help with that a lot. But for what appears to be the common use case of PerspectiveAPI, filtering a comments section, the effects are different. The difference in score between "Hello, my name is Brad" and "Hello, my name is <UNK>" may in many cases cross the threshold from "auto-accept" to "human review". Is every Discourse user of PerspectiveAPI really going through their entire "human review" queue and making an informed decision about each one? Or are some of them falling through the cracks and being implicitly rejected? I forget to review comments on my blog sometimes, and it's the usual low-volume setup where _every_ comment is reviewed.

It is a problem if people have a more frustrating commenting experience just because their name is what PerspectiveAPI considers an unknown word, or because they mentioned LGBT or Islam, while people who use names and self-identifications that sound like fine upstanding Wikipedians (I'm sure you don't need me to point out all the correlates of this) get auto-approved, possibly even while saying something nasty.

One example that comes to mind of how consumers of PerspectiveAPI are not sufficiently warned is [Wired's "Internet Troll Map"](https://www.wired.com/2017/08/internet-troll-map/). This is not a map of Internet trolling. This is a map of racial diversity plus an outlier in Vermont. And the author and data visualizer of that article had no idea that that's what they were making.

But how to properly inform API consumers about what they're using is a bit off-topic. To return the focus to the topic of this issue report:

This model thinks "Islam" is very toxic and "Christianity" is not, that "Mexican" is possibly toxic and "Italian" is not, that "I am a gay woman" is possibly toxic and "I am a straight man" is not, and that "My name is <_unknown word_>" is possibly toxic and "My name is <_name correlated with privilege in America_>" is not. That's very bad. I'm asking you and Google to put a higher priority on fixing it.

**Comment by iislucas** [comment]
Thanks, happy to discuss!

We initially also shared your concern about inappropriate usage, which is why we started (and still have) an [application](http://perspectiveapi.com/#/) and whitelist process for using the API. But so far, we've been  impressed by the thoughtfulness of users of the API: Disqus only provides scores to sort, not to auto-reject. [The Coral Project's post](https://blog.coralproject.net/toxic-avenging/) is another good example of how comment platforms are using Perspective API thoughtfully and not for auto-rejection. Discourse's approach to implementation is like the others also impressively thoughtful: https://meta.discourse.org/t/evaluating-googles-perspective-api-on-your-discourse-forum/79971  (Although I haven't seen much actual use in Discourse yet, probably because we are whitelist only). I don't know of anyone doing auto-reject as you suggest, and as I said I when we discuss it with users, we advise against use for auto-rejection. So I think that this approach is actually working well, and that your fear, which we initially shared, is not born out by real usage. 

In [Wired's Troll Map article](https://www.wired.com/2017/08/internet-troll-map/) they say they use a threshold of 0.9: we haven't seen any unintended biases' affecting the models at that threshold. They also provide some typical examples of comments. You could also look at the wikipedia data if you through the model on real comments was biased in some way (it doesn't look like it to me). So overall, I think your claim about it being a map of racial diversity is wrong; the map also doesn't have a close correlation to racial diversity on census data either. What the map means, I don't really know - but the interesting thing to me in the visualization is the clock: if it was a map of something not correlated with local time, then you also wouldn't see that results you do. More generally, and somewhat off-topic: I actually think that analysis at scale is one of the more interesting use-cases of ML in this space. We wrote a paper published at WWW on how to do that and cross check for confounding variables here: https://arxiv.org/abs/1610.08914 

You seem to be mistaken about the scores for "Islam", "Mexican", "I am a gay woman", and the other examples; they are all "Unlikely to be perceived as toxic" according to the toxicity model and severe toxicity models (you can test the toxicity model, which is much more sensitive than severe toxicity, here: http://perspectiveapi.com/). 

With regard to your initial point, I think exploring ablation of names is an interesting idea w.r.t. mitigating and measuring name-related biases. I'll file a separate issue to explore that, as I think it's not really related to 90% of the discussion in this thread now :)
